lib/ts/controllers/s-modal.ts

```diff
@@ -0,0 +1,88 @@
+class ModalController extends Stacks.StacksController {
+    static targets = ["modal"];
+
+    // TODO should Stimulus bind this? Needed to keep TS error-free
+    private modalTarget!: HTMLElement;
+
+    private _boundClickFn!: any;
+    private _boundKeypressFn!: any;
+
+    connect () { };
+
+    toggle (e: Event) {
+        e.preventDefault();
+        this._toggle();
+    }
+
+    _toggle (show?: boolean | undefined) {
+        var toShow = show;
+
+        // if we're letting the class toggle, we need to figure out if the popover is visible manually
+        if (typeof toShow === "undefined") {
+            toShow = this.modalTarget.getAttribute("aria-hidden") === "true";
+        }
+
+        // show/hide events trigger before toggling the class
+        this.triggerEvent(toShow ? "show" : "hide");
+
+        this.modalTarget.setAttribute("aria-hidden", toShow ? "false" : "true");
+        this.modalTarget.setAttribute("tab-index", toShow ? "0" : "-1");
+
+        if (toShow) {
+            this._bindDocumentEvents();
+        }
+        else {
+            this._unbindDocumentEvents();
+        }
+
+        // shown/hidden events trigger after toggling the class
+        this.triggerEvent(toShow ? "shown" : "hidden");
```

Comment:
99% of my use-cases will involve removing the modal element from the DOM after hiding it, so I'd like the `shown` and `hidden` events to fire _after_ the animation completes so I can discard it appropriately.

Something along the lines of the following could help facilitate that:

```
connect() {
    this.element.addEventListener('transitionend', () => {
        this.triggerEvent(this.element.getAttribute('aria-hidden') === 'false' ? 'shown' : 'hidden');
    });
}
```

Reply:
Looks like many browsers don't have full support for this, most notably Chome: https://caniuse.com/#search=transitionend

Bootstrap ([for example](https://github.com/twbs/bootstrap/blob/master/js/src/util/index.js#L92)) deals with this by detecting whether the event was fired or not by setting a timeout (in js) that is exactly as long as the css transition length. I don't have the bandwidth right now to chase this down, but I'd be happy to accept a PR.